src/ProtoRoute/Message.hs

```diff
@@ -1,46 +1,154 @@
+                       WrongName  String |
+                       WrongType  String |
+                       Invalid    String deriving (Show, Eq)
+
+incomplete :: ValidationError
+incomplete = Incomplete "Required field not filled"
+
+incorrectFieldNames :: ValidationError
+incorrectFieldNames = WrongName "Wrong field names"
+
+incorrectMsgNames :: ValidationError
+incorrectMsgNames = WrongName "Message names do not match"
+
+incorrectTypes :: ValidationError
+incorrectTypes = WrongType "There are mismatching types"
 
-data Field = Field
-    { fieldName :: FieldName
-    , fieldValue :: FieldValue
-    } deriving (Show)
+invalid :: ValidationError
+invalid = Invalid "Your message is invalid"
 
+-- Shows value of a message field
 fromFV :: FieldValue -> String
 fromFV (FText txt) = fromTV txt
 fromFV (FInt num)  = fromTV num
 fromFV (FMsg msg)  = fromTV msg
 
+-- Shows value inside FieldValue constructor
 fromTV :: (Show a) => TValue a -> String
 fromTV (Req val)  = show val
 fromTV (Opt txt)  = show txt
 fromTV (Rep list) = show list
 fromTV Error = "Error"
 
+-- Checks if rules in msg match those in corresponding schema
+correct :: TValue a -> SchemaRule -> Bool
+correct (Req _) SReq = True
+correct (Opt _) SOpt = True
+correct (Rep _) SRep = True
+correct _ _ = False
+
+-- Checks if types are correct
+okTypes :: FieldValue -> SchemaRule -> SchemaType -> Bool
+okTypes (FText tvA) rA fTA = fTA == SText && correct tvA rA
+okTypes (FInt tvB) rB fTB  = fTB == SInt && correct tvB rB
+okTypes (FMsg tvC) rC fTC  = fTC == SMsg && correct tvC rC
+
+-- Verifies if there exists a validation error
+verify :: Bool -> ValidationError -> Either ValidationError ()
+verify True someError = Left someError
+verify False _ = Right ()
+
+-- Checks if a required field is filled
+reqValThere :: SchemaField -> MessageField -> Bool
+reqValThere (SchemaField SReq _ sfn) mf =
+    case mf of
+        (MessageField sfn _) -> True
+reqValThere _ _ = False
+
+-- Validates single field, checking msg attempt against schema
+validateField :: MessageField -> SchemaField -> Either ValidationError ()
+validateField (MessageField mfn mfv) (SchemaField sfr sft sfn) =
+    case wrongFieldNames of
+        Left se  -> Left se
+        Right () ->
+          case wrongType of
+              Left se  -> Left se
+              Right () ->
+                case noReqVal of
+                    Left se  -> Left se
+                    Right () -> Right ()
+  where
+    wrongFieldNames = verify (mfn /= sfn) incorrectFieldNames
+    wrongType = verify (not $ okTypes mfv sfr sft) incorrectTypes
+    noReqVal = verify (not $ reqValThere (SchemaField sfr sft sfn) (MessageField mfn mfv))
+                      incomplete
+
+validateMessage :: Message -> MessageSchema -> Either ValidationError ()
+validateMessage (Message mn mFields) (MessageSchema sn sFields) = do
+    let wrongMessageNames = verify (mn /= sn) incorrectMsgNames
+    case wrongMessageNames of
+        Left se  -> Left se
```

Comment:
This is exactly what `Either` `Monad` instance for. Instead of pattern-matching on `Either` you can just write:

```haskell
validateMessage ... = do
    verify (mn /= sn) incorrectMsgNames
    let errors = ...
```

It's easier to see after you rewrite `do`-notation into `>>=` syntax:

```haskell
verify foo bar >>= \() -> let errors = ...
````